feat(panes/nav): persist pane state; rename side panel to lesson

no-stack-dub-sack opened the pull request:
<!-- freeCodeCamp Pull Request Template -->

<!-- IMPORTANT Please review https://github.com/freeCodeCamp/freeCodeCamp/blob/staging/CONTRIBUTING.md for detailed contributing guidelines -->
<!-- Help with PRs can be found at https://gitter.im/FreeCodeCamp/Contributors -->
<!-- Make sure that your PR is not a duplicate -->

#### Pre-Submission Checklist
<!-- Go over all points below, and after creating the PR, tick all the checkboxes that apply. -->
<!-- All points should be verified, otherwise, read the CONTRIBUTING guidelines from above-->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] Your pull request targets the `staging` branch of freeCodeCamp.
- [ ] Branch starts with either `fix/`, `feature/`, or `translate/` (e.g. `fix/signin-issue`)
- [x] You have only one commit (if not, [squash](http://forum.freecodecamp.org/t/how-to-squash-multiple-commits-into-one-with-git/13231) them into one commit).
- [x] All new and existing tests pass the command `npm test`. Use `git commit --amend` to amend any fixes.

#### Type of Change
<!-- What type of change does your code introduce? After creating the PR, tick the checkboxes that apply. -->
- [ ] Small bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds new functionality)
- [ ] Breaking change (fix or feature that would change existing functionality)
- [ ] Add new translation (feature adding new translations)

#### Checklist:
<!-- Go over all points below, and after creating the PR, tick the checkboxes that apply. -->
<!-- If you're unsure about any of these, don't hesitate to ask in the Contributors room linked above. We're here to help! -->
- [x] Tested changes locally.
- [x] Addressed currently open issue (replace XXXXX with an issue no in next line)

Closes #16407 
Closes #16453

Closes #16492 

#### Description
<!-- Describe your changes in detail -->
@BerkeleyTrue this implements tracking pane state (panes opened/closed, and position) per your suggestion in #16407 

**EDIT: RESOLVED.** 

~~however, there are a few issues and this is a work in progress. I discovered that in some cases, with certain panes closed (e.g. map & preview), when the next challenge loads, an unnecessary vertical divider would appear (you can drag it, but it doesn't resize anything).~~

~~I also thought that persisting pane width might be good, and this appeared to remove the extraneous divider in that instance, however it raised some other issues.~~

~~Since "Map" pane already has state by the time "Preview", "Side Panel" and "index" load, map already has a value of `50` which effectively hides the "Side Panel". **Edit:** see comment below for possible solution to one issue.~~

~~Any suggestions regarding the extraneous dividers?~~

no-stack-dub-sack committed:
feat(panes/nav): persist pane state across transitions; rename side panel to lesson

no-stack-dub-sack commented on common/app/Panes/redux/index.js:
```diff
@@ -176,14 +171,22 @@ export default function createPanesAspects({ createPanesMap }) {
           panesMap,
           panes,
-          panesByName: panes.reduce((panes, { name }, index) => {
+          panesByName: panes.reduce((__panes, { name }, index) => {
```
I can change this back from `__panes` to `panes` - I changed it at one point to try something, but no longer need it. Forgot about this.

no-stack-dub-sack commented on common/app/Panes/redux/index.js:
```diff
@@ -176,14 +171,22 @@ export default function createPanesAspects({ createPanesMap }) {
+                state,
+                ['panesByName', name, 'dividerLeft'],
+                dividerLeft
+              ),
```
instead of this, perhaps I could add this before `panesMap` is defined (above line 167): 
```js
const wasStepChallenge = _.findIndex(state.panesMap, {name: ‘Step’}) !== -1;
```
this, right after `panesMap` is defined (line 168):
```js
const isStepChallenge = _.findIndex(panesMap, {name: ‘Step’}) !== -1;
```
and then replace lines 176-182 above with:
```js
dividerLeft: name === ‘Map’ && wasStepChallenge && !isStepChallenge 
  ? 25 
  : _.get(
    state,
    ['panesByName', name, 'dividerLeft'],
    dividerLeft
 ),
```
This should prevent "Map"'s `dividerLeft` from defaulting to `50` and covering the side panel when moving from a Map/Step challenge/intro. Haven't tested this out yet though.

BerkeleyTrue replied:
This approach tightly couples Panes to the content. We should avoid this.

What about iterating through the old and new panes left to right and assigning the previous value of dividerLeft to the new pane. Then, once a pane is detected that did not exist before, we split the remaining area to the remaining panes.

no-stack-dub-sack replied:
@BerkeleyTrue Yes, that could work, but the result would be less balanced visually, because when moving from Step/Map to 4 panes, we'd have Map taking up 50% of the view width and the remaining panes ~16% each. Let me think about this, I might have another solution that wouldn't require any addition utils to iterate over the panes and would still keep the panes and content decoupled.

BerkeleyTrue replied:
Yeah that could be a problem. Let me know what you come up with.

no-stack-dub-sack commented on common/app/Panes/redux/index.js:
```diff
@@ -172,19 +166,24 @@ export default function createPanesAspects({ createPanesMap }) {
+                return panes;
+              },
+            {}
+          )
```
@BerkeleyTrue Ok, I think this does it. We can just compare prev panes to new panes, and base the rest on that: if the number or type of panes that exist between transitions doesn't change, don't change or recalculate anything &mdash; just persist `state.panesByName` which preserves all `dividerLeft` values as well as `isHidden`. 

Otherwise, just recalculate `dividerLeft` and set `isHidden` to false for every pane as normal. 

The latter should only happen when new types of panes are introduced (e.g. from Map/Step to challenge), but doesn't couple the state with any type of pane or content in particular.

no-stack-dub-sack commented on common/app/Panes/redux/index.js:
```diff
@@ -202,7 +201,7 @@ export default function createPanesAspects({ createPanesMap }) {
         }, 0);
         let numOfHidden = 0;
-        return {
+        return numOfPanes < 1 ? state : {
```
This is to prevent a single pane from being toggled off, when only one pane is being rendered, so the user never ends up with a blank screen.

BerkeleyTrue replied:
We have an issue and incoming PR for this will render a hidden message.

no-stack-dub-sack commented on common/app/Nav/Nav.jsx:
```diff
@@ -51,7 +51,7 @@ const mapStateToProps = createSelector(
       panes: panes.map(({ name, type }) => {
         return {
-          content: name,
+          content: name === 'index' ? 'Editor' : name,
```
I also noticed that Modern challenges are the only ones where the side panel is actually called "Side Panel" and the editor is called "index" (in the Nav). Side Panel looks like it can be changed to "Lesson" with no side effects, but since the editor pane is coupled to the file key, we have to make the change to the Nav UI on the component level without affecting the pane name (short of changing the file key to "Editor" in all the seed code). So this is a little inelegant and couples the name determination to this specific component, but gives us nice nav consistency across the entire app. Does this work for you?

BerkeleyTrue replied:
This is by design and there is already an issue for it. The name comes from the challenge, so creators can call the file what ever they want. 

The Side Panel can be changed in the paneMap definition itself

no-stack-dub-sack commented on common/app/routes/Challenges/views/Modern/Show.jsx:
```diff
@@ -50,7 +50,7 @@ export const mapStateToPanes = addNS(
       }, {
         [types.toggleMap]: 'Map',
-        [types.toggleSidePanel]: 'Side Panel'
+        [types.toggleSidePanel]: 'Lesson'
```
See note at top.

BerkeleyTrue replied:
Works

no-stack-dub-sack commented:
@BerkeleyTrue Ok, I removed the other pieces, and just left the code that addresses the original issue and renamed Side Panel. Are you good with the new approach for maintaining pane state?

BerkeleyTrue commented:
> ...just for consistency 

The original intent here was to replicate an actual editor, with the name of the file you are actually editing at the top of the file but I see the issue with the name in the nav that could confuse new users. The problem with renaming it to editor is when you have more than one,  you can't call them both `Editor`.

A compromise: When there is only one editor we can default to `Editor` and when there is more than one we could name them by file. Thoughts?

BerkeleyTrue commented:
merged in 62e079c7215fc13296e934ff304b6894e4a45dca

I updated the logic so that if the panes are equal we do not modify state. 

Thanks again @no-stack-dub-sack . Let me know what you think of my comment above. 

Happy Coding

no-stack-dub-sack commented:
@BerkeleyTrue Ahh... I get it now. I didn't realize that was the original intent. Do we plan on getting to a place where there will be multiple editor panes for diff langs? HTML/CSS/JS so projects can be done right on fCC? I know that was the original hope, but wasn't sure if that was still on the table. 

I think your suggested compromise works, also, maybe just calling it index.js or index.jsx would also clear up the confusion. Thoughts on that?

BerkeleyTrue commented:
> ...multiple editor panes for diff langs?

Yeah, or multi editors for the same lang to do lessons on import/export/require/modules. The UI is there but the build logic is not complete/tested. Right now launching beta is priority and we don't need multi for that.

> maybe just calling it index.js or index.jsx

Yeah, that was my intent but the multi file system was built way before panes was complete so there was a disconnect there. Panes uses file.name which does not have the extension. This is a bug.

no-stack-dub-sack commented:
@BerkeleyTrue No worries then, I think in the meantime your first suggestion will do then. Should we open a new issue for this?

> When there is only one editor we can default to Editor and when there is more than one we could name them by file.

BerkeleyTrue commented:
Sure, do you mind opening it?
